src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py

```diff
@@ -276,156 +291,199 @@ def __get_test_list_by_runtime(self):
-            logger.info(f'Test count from cache: {len(cached_test_list)}')
-            logger.info(f'Test count from runtime: {len(runtime_test_test)}')
-        return cached_test_list, runtime_test_test
+            logger.info(f'Test count from cache: {len(cached_test_dict)}')
+            logger.info(f'Test count from runtime: {len(runtime_test_dict)}')
+        return cached_test_dict, runtime_test_dict
 
-    def __prepare_smart_filters(self, proved_test_list:list):
+    def __prepare_smart_filters(self, proved_test_dict: dict):
         res_test_filters = list()
         def_length = len(self._command) + len(" --gtest_filter=")
-        idx = len(proved_test_list)
-        for i in range(len(proved_test_list)):
-            if proved_test_list[i]._time == -1:
+
+        idx = len(proved_test_dict)
+        for i in range(len(proved_test_dict)):
+            if proved_test_dict[list(proved_test_dict)[i]] == -1:
                 idx = i
                 break
-        
+
         # Run crashed tests in a separed thread
-        if idx < len(proved_test_list):
-            while proved_test_list[idx]._time == -1:
-                test = proved_test_list.pop(idx)
-                res_test_filters.append(test._name)
-                if idx >= len(proved_test_list):
+        if idx < len(proved_test_dict):
+            while list(proved_test_dict.values())[idx] == -1:
+                res_test_filters.append(list(proved_test_dict.keys())[idx] + "*:")
+                proved_test_dict.pop(list(proved_test_dict.keys())[idx])
+                if idx >= len(proved_test_dict):
                     break
 
         longest_device = ""
         for device in self._available_devices:
             if len(device) > len(longest_device):
                 longest_device = device
 
-        # prepare gtest filters per worker according command line length limitation
-        while len(proved_test_list) > 0:
-            test_times = []
-            is_not_full = True
-            worker_test_filters = list()
+        real_worker_num = self._worker_num * len(self._available_devices)
+        workers = [(0, "")] * real_worker_num
+
+        tests_sorted = sorted(proved_test_dict.items(), key=lambda i: i[1], reverse=True)
+        for test_item in tests_sorted :
+            test_pattern = f'"{self.__replace_restricted_symbols(test_item[0])}"*:'
+            test_time = test_item[1]
+            w_id = -1
+            for i in range(len(workers)) :
+                if len(workers[i][1]) + def_length + len(test_pattern.replace(self._device, longest_device)) < MAX_LENGHT :
+                    w_id = i
+                    break
 
-            real_worker_num = self._worker_num * len(self._available_devices)
+            if w_id == -1 :
+                workers.append((test_time, test_pattern))
+            else :
+                workers[w_id] = (workers[w_id][0] + test_time, workers[w_id][1] + test_pattern)
+            workers.sort()
```

Comment:
sort many times...
Let's use `multimap` -> {time, filters}. The code will be more clear and easy to understand :)

and do not forget about `is_full` flag to avoid overflow of command-line  string...